From 979bfcaf1727d301ac43fd0c75c367f591290456 Mon Sep 17 00:00:00 2001
From: Biswajit Paul <biswajitpaul@codeaurora.org>
Date: Wed, 27 Jul 2016 17:01:06 -0700
Subject: [PATCH 1/9] sepolicy : Allow qti_debugfs_domain permission to access
 debugfs

Whiltelist qti_debugfs_domain to access debugfs

CRs-Fixed: 1025226
Change-Id: I1f9deb0df3759ecbe2ffd99330ff8a3662648765
---
 domain.te | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/domain.te b/domain.te
index 45569de..f66c33b 100644
--- a/domain.te
+++ b/domain.te
@@ -570,7 +570,7 @@ neverallow * domain:file { execute execute_no_trans entrypoint };
 # Instead, if access to part of debugfs is desired, it should have a
 # more specific label.
 # TODO: fix system_server and dumpstate
-neverallow { domain -init -system_server -dumpstate } debugfs:file no_rw_file_perms;
+neverallow { domain -init -system_server -dumpstate userdebug_or_eng(`-qti_debugfs_domain')} debugfs:file no_rw_file_perms;
 
 neverallow {
   domain
-- 
2.7.4


From 794c36a5d6f083246811875259e3062d182ffa1c Mon Sep 17 00:00:00 2001
From: maxwen <max.weninger@gmail.com>
Date: Tue, 4 Oct 2016 13:25:07 +0200
Subject: [PATCH 2/9] sepolicy: add qti_debugfs_domain attribute

Conflicts:
	attributes

Change-Id: I90a501a7db1a3fae1f66f9d0ccf3cf8db236799f
---
 attributes | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/attributes b/attributes
index a846c34..72a7bad 100644
--- a/attributes
+++ b/attributes
@@ -113,3 +113,6 @@ attribute boot_control_hal;
 # postinstall. This includes the background daemon and the sideload tool from
 # recovery for A/B devices.
 attribute update_engine_common;
+
+# Domain type used for debugfs access
+attribute qti_debugfs_domain;
-- 
2.7.4


From 159f895ace87346c7e3305c923fca3cb8b641e3c Mon Sep 17 00:00:00 2001
From: Hashcode <hashcode0f@gmail.com>
Date: Tue, 13 Dec 2016 15:51:16 -0800
Subject: [PATCH 3/9] sepolicy: fix denials for rootfs:lnk_file

Since android-7.1.1 I'm seeing denials for any domain accessing the
/vendor symlink.  Apparently related to getattr.  Not sure which
commit caused the denials, but we are probably ok allowing getattr
on rootfs:lnk_file.  The commit adds it.

sample dmesg lines:
type=1400 audit(1481572084.260:4): avc: denied { getattr } for pid=294 comm="sysinit" path="/vendor" dev="rootfs" ino=5603 scontext=u:r:sysinit:s0 tcontext=u:object_r:rootfs:s0 tclass=lnk_file permissive=0
type=1400 audit(1481572084.680:5): avc: denied { getattr } for pid=379 comm="90userinit" path="/vendor" dev="rootfs" ino=5603 scontext=u:r:sysinit:s0 tcontext=u:object_r:rootfs:s0 tclass=lnk_file permissive=0
type=1400 audit(1481572084.860:6): avc: denied { getattr } for pid=493 comm="wan_start" path="/vendor" dev="rootfs" ino=5603 scontext=u:r:startril:s0 tcontext=u:object_r:rootfs:s0 tclass=lnk_file permissive=0
type=1400 audit(1481572084.890:7): avc: denied { getattr } for pid=506 comm="audioserver" path="/vendor" dev="rootfs" ino=5603 scontext=u:r:audioserver:s0 tcontext=u:object_r:rootfs:s0 tclass=lnk_file permissive=0
type=1400 audit(1481572084.900:8): avc: denied { getattr } for pid=510 comm="mediacodec" path="/vendor" dev="rootfs" ino=5603 scontext=u:r:mediacodec:s0 tcontext=u:object_r:rootfs:s0 tclass=lnk_file permissive=0
type=1400 audit(1481572084.970:9): avc: denied { getattr } for pid=511 comm="mediadrmserver" path="/vendor" dev="rootfs" ino=5603 scontext=u:r:mediadrmserver:s0 tcontext=u:object_r:rootfs:s0 tclass=lnk_file permissive=0
type=1400 audit(1481572085.070:10): avc: denied { getattr } for pid=543 comm="ath6kl-service" path="/vendor" dev="rootfs" ino=5603 scontext=u:r:ath6kl-service:s0 tcontext=u:object_r:rootfs:s0 tclass=lnk_file permissive=0
type=1400 audit(1481572084.770:11): avc: denied { getattr } for pid=829 comm="bootanimation" path="/vendor" dev="rootfs" ino=5603 scontext=u:r:bootanim:s0 tcontext=u:object_r:rootfs:s0 tclass=lnk_file permissive=0

Change-Id: Icfafadf18cbe29e3ca0316149d97e6ea24274941
---
 domain.te | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/domain.te b/domain.te
index f66c33b..a2bb7a2 100644
--- a/domain.te
+++ b/domain.te
@@ -62,7 +62,7 @@ allow domain debuggerd:unix_stream_socket connectto;
 
 # Root fs.
 allow domain rootfs:dir search;
-allow domain rootfs:lnk_file read;
+allow domain rootfs:lnk_file { read getattr };
 
 # Device accesses.
 allow domain device:dir search;
-- 
2.7.4


From 241f47adf22428069630ab4dd14b7cbfe5718cd7 Mon Sep 17 00:00:00 2001
From: kldoc <fred.berbigier@gmail.com>
Date: Wed, 5 Oct 2016 10:42:20 +0800
Subject: [PATCH 4/9] allow-execmod-for-cameraserver

Needed for camera to work on some devices

Change-Id: I9264970f749aa166cdb1f36a9e4d813c30c31933
---
 camera.te | 2 ++
 domain.te | 5 +++--
 2 files changed, 5 insertions(+), 2 deletions(-)
 create mode 100644 camera.te

diff --git a/camera.te b/camera.te
new file mode 100644
index 0000000..2e2f519
--- /dev/null
+++ b/camera.te
@@ -0,0 +1,2 @@
+# camera
+type camera, domain;
diff --git a/domain.te b/domain.te
index a2bb7a2..9c7fa00 100644
--- a/domain.te
+++ b/domain.te
@@ -434,6 +434,7 @@ neverallow * {
   -apk_data_file
   -app_data_file
   -asec_public_file
+  -system_file
 }:file execmod;
 
 # Do not allow making the stack or heap executable.
@@ -443,8 +444,8 @@ neverallow * self:process { execstack execheap };
 
 # prohibit non-zygote spawned processes from using shared libraries
 # with text relocations. b/20013628 .
-neverallow { domain -appdomain } file_type:file execmod;
-
+#camera_server neverallow needed for Android N
+neverallow { domain -appdomain userdebug_or_eng(`-camera -cameraserver')} file_type:file execmod;
 neverallow { domain -init } proc:{ file dir } mounton;
 
 # Ensure that all types assigned to processes are included
-- 
2.7.4


From 686f3a445ad58b558bfbb8b8b003e5cd1d305d45 Mon Sep 17 00:00:00 2001
From: Marko Man <darkobas@gmail.com>
Date: Thu, 20 Oct 2016 01:37:46 +0200
Subject: [PATCH 5/9] sepolicy: allow fsck_untrusted

Change-Id: Ia3a2c6a3f37b97910b5c62488c32818ebf1cc980
---
 domain.te | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/domain.te b/domain.te
index 9c7fa00..cf9efc3 100644
--- a/domain.te
+++ b/domain.te
@@ -579,6 +579,7 @@ neverallow {
   -recovery
   -sdcardd
   -vold
+  -fsck_untrusted
 } fuse_device:chr_file open;
 neverallow {
   domain
@@ -590,6 +591,7 @@ neverallow {
   -system_server
   -ueventd
   -vold
+  -fsck_untrusted
 } fuse_device:chr_file *;
 
 # Profiles contain untrusted data and profman parses that. We should only run
-- 
2.7.4


From 29ab54f51c75ea5183f786a19c674e65bcde7f59 Mon Sep 17 00:00:00 2001
From: LuK1337 <priv.luk@gmail.com>
Date: Tue, 25 Oct 2016 23:26:12 +0200
Subject: [PATCH 6/9] Add rules required for TARGET_HAS_LEGACY_CAMERA_HAL1

Change-Id: Ica4926fa3acf88b95c687f25eec4dbc68ea41fbb
---
 Android.mk     | 1 +
 mediaserver.te | 6 ++++++
 2 files changed, 7 insertions(+)

diff --git a/Android.mk b/Android.mk
index 0bfa54d..aa28ffb 100644
--- a/Android.mk
+++ b/Android.mk
@@ -95,6 +95,7 @@ $(sepolicy_policy.conf): $(call build_policy, $(sepolicy_build_files))
 	$(hide) m4 $(PRIVATE_ADDITIONAL_M4DEFS) \
 		-D mls_num_sens=$(PRIVATE_MLS_SENS) -D mls_num_cats=$(PRIVATE_MLS_CATS) \
 		-D target_build_variant=$(TARGET_BUILD_VARIANT) \
+		-D target_has_legacy_camera_hal1=$(TARGET_HAS_LEGACY_CAMERA_HAL1) \
 		-s $^ > $@
 	$(hide) sed '/dontaudit/d' $@ > $@.dontaudit
 
diff --git a/mediaserver.te b/mediaserver.te
index 5fbaa30..dc05e14 100644
--- a/mediaserver.te
+++ b/mediaserver.te
@@ -94,6 +94,12 @@ allow mediaserver processinfo_service:service_manager find;
 allow mediaserver scheduling_policy_service:service_manager find;
 allow mediaserver surfaceflinger_service:service_manager find;
 
+ifelse(target_has_legacy_camera_hal1, `true',
+  allow mediaserver cameraproxy_service:service_manager find;
+  allow mediaserver cameraserver_service:service_manager add;
+,
+)
+
 # /oem access
 allow mediaserver oemfs:dir search;
 allow mediaserver oemfs:file r_file_perms;
-- 
2.7.4


From b996f0892f06eed8ce4cdd0d8f46b187bef03179 Mon Sep 17 00:00:00 2001
From: Ricardo Cerqueira <ricardo@cyngn.com>
Date: Thu, 6 Nov 2014 19:16:37 +0000
Subject: [PATCH 7/9] domain: Create a dummy domain for qc's rmt and let it
 access /dev/mem

Old (pre-L) binaries will fail hard if unable to access mem. Create
a domain that rmt_storage can use to get those perms

Limit to userdebug and eng builds

Change-Id: I3ff72b11434df7217631330370b05f59a13348e9
---
 domain.te | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/domain.te b/domain.te
index cf9efc3..32834c4 100644
--- a/domain.te
+++ b/domain.te
@@ -170,6 +170,7 @@ neverallow {
   -vold
 } self:capability mknod;
 
+userdebug_or_eng(`attribute rmt_placeholder;')
 # Limit raw I/O to these whitelisted domains. Do not apply to debug builds.
 neverallow {
   domain
@@ -181,6 +182,7 @@ neverallow {
   -healthd
   -uncrypt
   -tee
+  userdebug_or_eng(`-rmt_placeholder')
 } self:capability sys_rawio;
 
 # No process can map low memory (< CONFIG_LSM_MMAP_MIN_ADDR).
@@ -237,8 +239,8 @@ neverallow { domain -init -system_server -ueventd } hw_random_device:chr_file *;
 neverallow * { file_type -exec_type -postinstall_file }:file entrypoint;
 
 # Ensure that nothing in userspace can access /dev/mem or /dev/kmem
-neverallow { domain -kernel -ueventd -init } kmem_device:chr_file *;
-neverallow * kmem_device:chr_file ~{ create relabelto unlink setattr };
+neverallow { domain userdebug_or_eng(`-rmt_placeholder') -kernel -ueventd -init } kmem_device:chr_file *;
+neverallow { domain userdebug_or_eng(`-rmt_placeholder') } kmem_device:chr_file ~{ create relabelto unlink setattr };
 
 # Only init should be able to configure kernel usermodehelpers or
 # security-sensitive proc settings.
-- 
2.7.4


From 4282468ea421fdee774a50915039bbfbe01a29bf Mon Sep 17 00:00:00 2001
From: Steve Kondik <steve@cyngn.com>
Date: Wed, 26 Oct 2016 20:34:49 -0700
Subject: [PATCH 8/9] sepolicy: Forward declaration of qti-testscripts domain

 * This is used by the QCOM-specific policy and needs to dodge
   a neverallow rule on debug builds for test purposes.
 * For non-QCOM devices which don't include device/qcom/sepolicy,
   we need to forward-declare the domain so the build completes.

Change-Id: Ie22e4c550d24ffb8bf7f9320751d41603eff93f5
---
 qti-testscripts.te | 3 +++
 1 file changed, 3 insertions(+)
 create mode 100644 qti-testscripts.te

diff --git a/qti-testscripts.te b/qti-testscripts.te
new file mode 100644
index 0000000..c616f98
--- /dev/null
+++ b/qti-testscripts.te
@@ -0,0 +1,3 @@
+userdebug_or_eng(`
+  type qti-testscripts, domain, domain_deprecated, mlstrustedsubject;
+')
-- 
2.7.4


From 369dade5d319cf4ab376f8675024bac4668800e9 Mon Sep 17 00:00:00 2001
From: Steve Kondik <steve@cyngn.com>
Date: Wed, 26 Oct 2016 02:42:18 -0700
Subject: [PATCH 9/9] sepolicy: Additional userdebug/eng restrictions for 7.1

Change-Id: I499603b55305606100e7b0bc188969583803e4f0
---
 domain.te | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/domain.te b/domain.te
index 32834c4..abf27b9 100644
--- a/domain.te
+++ b/domain.te
@@ -492,7 +492,7 @@ neverallow {
 
 # Only domains spawned from zygote and runas may have the appdomain attribute.
 neverallow { domain -runas -zygote } {
-  appdomain -shell userdebug_or_eng(`-su') -bluetooth
+  appdomain -shell userdebug_or_eng(`-su -sudaemon -qti-testscripts') -bluetooth
 }:process { transition dyntransition };
 
 # Minimize read access to shell- or app-writable symlinks.
-- 
2.7.4

